core/include/drivers/clk.h

```diff
@@ -0,0 +1,287 @@
+/* SPDX-License-Identifier: BSD-2-Clause */
+/*
+ * Copyright (c) 2021, Bootlin
+ */
+
+#ifndef CLK_H
+#define CLK_H
+
+#include <kernel/dt.h>
+#include <stdint.h>
+#include <sys/queue.h>
+
+#define CLK_NO_SCMI_ID	-1
+
+struct clk;
+
+/**
+ * struct clk_ops
+ *
+ * @enable: Enable the clock
+ * @disable: Disable the clock
+ * @is_enabled: Check if the clock is enabled
+ * @set_parent: Set the clock parent based on index
+ * @get_parent: Get the current parent index of the clock
+ * @set_rate: Set the clock rate
+ * @get_rate: Get the clock rate
+ * @get_rates_array: Get the supported clock rates as array
+ */
+struct clk_ops {
+	int (*enable)(struct clk *clk);
+	void (*disable)(struct clk *clk);
+	int (*is_enabled)(struct clk *clk);
+	int (*set_parent)(struct clk *clk, uint8_t index);
+	uint8_t (*get_parent)(struct clk *clk);
+
+	int (*set_rate)(struct clk *clk, unsigned long rate,
+			unsigned long parent_rate);
+	unsigned long (*get_rate)(struct clk *clk,
+				  unsigned long parent_rate);
+	int (*get_rates_array)(struct clk *clk, size_t start_index,
+			       unsigned long *rates, size_t *nb_elts);
+};
+
+/* Flags for clock */
+#define CLK_SET_RATE_GATE	BIT(0) /* must be gated across rate change */
+#define CLK_SET_PARENT_GATE	BIT(1) /* must be gated across re-parent */
+#define CLK_SET_RATE_PARENT	BIT(2) /* propagate rate change up one level */
+
+/**
+ * struct clk - Clock structure
+ *
+ * @name: Clock name
+ * @priv: private data for the clock provider
+ * @ops: Clock operations
+ * @parents: Array of parents of the clock
+ * @parent: Current parent
+ * @num_parents: Number of parents
+ * @rate: Current clock rate (cached after init or rate change)
+ * @scmi_id: SCMI identifier for the clock (used by the generic SCMI clock
+ *	     support)
+ * @scmi_channel_id: SCMI channel for the clock (used by the generic SCMI clock
+ *	     support)
```

Comment:
these should be outside the clock frameork I think.
SCMI clocks should rely on this clock framework, as a clock consumer like any other clock consumer.

Reply:
Actually, the clock provider set the SCMI ID by itself. I'm not sure who will set the clocks ID if not the provider.
How will the matching be done between hardware clocks and SCMI clocks ID then ?
Currently, the use case is that the clock provider allocate the clock and then sets the SCMI ID to match the hardware clock. But I agree this is not ideal.

Reply:
After thinking a bit more about that, I could add an interface to translate the equivalent of devicetree clock-cells to a struct *clk. The generic scmi clk layer will then export a function to match the clk struct to a scmi_id.
The flow for the platform code would be the following:

``` C
/* Provided by generic clock support */
struct clock_args args = {PMC_TYPE_PERIPHERAL, 42};
clk = clk_get_from_provider("provider_name", args);

/* Provided by generic scmi clk support */
clk_scmi_match(clk, channel_id, scmi_id);
```

Does it sound good to you ?

Reply:
I think scmi references should move out of the generic clock framework. If an SCMI server exposes clocks, it should access the clock as any other clock consumer in optee, without specific hooks in the clock framework.

Binding platform clock to SCMI clock is platform specific. What about using a specific node in the DT for defining the clocks exposed, and a dedicated clock consumer driver to exposes these clock to the SCMI server? For example, assuming channel 0 exposes 2 SCMI clocks:
```
scmi-clock-ch0 {
	compatible = "scmi-clock-consumer";
	scmi-channel-id = <0>;

	scmi-clock-ch0@0 {
		reg = <0>;      // the scmi clock ID used to expose the clock
		clocks = <&some_clock>;   // reference to a platform clock
	}
	scmi-clock-ch0@1 {
		reg = <1>;      // the scmi clock ID used to expose the clock
		clocks = <&some_clock_phandle 12>; // reference to another platform clock
	}
}
```

Reply:
That would be a really clean solution ! I like this proposal which is way more flexible than what I have. I'll probably go this way.
Regarding the SCMI part, I did split it completely from the generic clock framework in my last revision.

Reply:
I'm not sure having a node per clock is a good ID, we explicitely switched away from that in the kernel. Maybe the mapping could be done using a single array of phandles as we know the SCMI IDs start at 0 and are contiguous.

Reply:
Having a single `clocks = <&clock1 arg1 arg 2 &clock2 arg3 arg4 ... >` is probably going to be difficult to read I guess and even harder to modify.
However, I agree that current proposal will yield a consequent number of nodes.

Reply:
On a specific SoC I'm working on, there is 80 clocks that needs to be exposed. Since the properties have the same name, it won't take a lot of space IMHO.
Your solution is more flexible regarding the per clock properties. For instance, allowed SCMI frequency could be specified in theses nodes.